<?xml version="1.0"?>
<robot name="robot">
    <link name="world"/>

    <link name="base">
        <inertial>
            <origin xyz="0 0 0.001" rpy="0 0 0"/>
            <mass value="2.0"/>
            <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
        </inertial>
        <visual>
            <origin xyz="0 0 0.001" rpy="0 0 0"/>
            <geometry>
                <cylinder radius="0.2" length="0.04"/>
            </geometry>
        </visual>
        <collision>
            <origin xyz="0 0 0.001" rpy="0 0 0"/>
            <geometry>
                <cylinder radius="0.2" length="0.04"/>
            </geometry>
        </collision>
    </link>

    <link name="arm1">
        <inertial>
            <origin xyz="0 0 0.50" rpy="0 0 0"/>
            <mass value="0.5"/> 
            <inertia ixx="0.04" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.04"/>
        </inertial>
        <visual>
            <origin xyz="0 0 0.50" rpy="0 0 0"/>
            <geometry>
                <cylinder radius="0.05" length="0.5"/>
            </geometry>
        </visual>
        <collision>
            <origin xyz="0 0 0.50" rpy="0 0 0"/>
            <geometry>
                <cylinder radius="0.05" length="0.5"/>
            </geometry>
        </collision>
    </link>

    <joint name="base-world" type="fixed">
        <parent link="world"/>
        <child link="base"/>
        <origin xyz="0 0 0"/>
    </joint>

    <joint name="joint_a" type="revolute">
        <axis xyz="0 0 1.0"/>
        <limit effort="160.00000" lower="-3.142" upper="3.142" velocity="-6.284"/>
        <parent link="base"/>
        <child link="arm1"/>
        <origin xyz="0 0 0.04"/>
    </joint>

    <link name="arm2">
        <inertial>
            <origin xyz="0.2 0 0" rpy="0 0 0"/>
            <mass value="0.4"/> 
            <inertia ixx="0.002" ixy="0.0" ixz="0.0" iyy="0.04" iyz="0.0" izz="0.04"/>
        </inertial>
        <visual>
            <origin xyz="0.2 0 0" rpy="0 3.14 0"/>
            <geometry>
                <cylinder radius="0.05" length="0.5"/>
            </geometry>
        </visual>
        <collision>
            <origin xyz="0.2 0 0" rpy="0 3.14 0"/>
            <geometry>
                <cylinder radius="0.05" length="0.5"/>
            </geometry>
        </collision>
    </link>
    <joint name="joint_b" type="revolute">
        <axis xyz="0.0 1.0 0.0"/>
        <limit effort="160.00000" lower="-3.142" upper="3.142" velocity="-6.284"/>
        <parent link="arm1"/>
        <child link="arm2"/>
        <origin xyz="0.0 0.2 0.45"/>
    </joint>

    <link name="arm3">
        <inertial>
            <origin xyz="0.05 0.20 0.0" rpy="0 0 0"/>
            <mass value="0.3"/> 
            <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.002" iyz="0.0" izz="0.04"/>
        </inertial>
        <visual>
            <origin xyz="0.05 0.20 0.0" rpy="3.14 3.14 0.0"/>
            <geometry>
                <cylinder radius="0.05" length="0.5"/>
            </geometry>
        </visual>
        <collision>
            <origin xyz="0.05 0.20 0.0" rpy="3.14 3.14 0.0"/>
            <geometry>
                <cylinder radius="0.05" length="0.5"/>
            </geometry>
        </collision>
    </link>
    <joint name="joint_c" type="revolute">
        <axis xyz="1.0 0.0 0.0"/>
        <limit effort="160.00000" lower="-0.78540" upper="3.142" velocity="-6.284"/>
        <parent link="arm2"/>
        <child link="arm3"/>
        <origin xyz="0.45 0.0 0.0"/>
    </joint>
    
    <ros2_control name="GazeboSystem" type="system">
    <hardware>
        <plugin>gazebo_ros2_control/GazeboSystem</plugin>
    </hardware>

    <joint name="joint_a">
        <command_interface name="position">
            <param name="min">-3.142</param>
            <param name="max">3.142</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
        <state_interface name="effort"/>
    </joint>

    <joint name="joint_b">
        <command_interface name="position">
            <param name="min">-3.142</param>
            <param name="max">3.142</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
        <state_interface name="effort"/>
    </joint>

    <joint name="joint_c">
        <command_interface name="position">
            <param name="min">-3.142</param>
            <param name="max">3.142</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
        <state_interface name="effort"/>
    </joint>
</ros2_control>

    <gazebo>
        <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
            <parameters>/home/aurovind/Documents/3arm_robotic_manipulator/DC1/config/robot_config.yaml</parameters>
        </plugin>
    </gazebo>
</robot>